tools/pythonpkg/src/vector_conversion.cpp

```diff
@@ -109,46 +117,127 @@ static string_t DecodePythonUnicode(T *codepoints, idx_t codepoint_count, Vector
 	return result;
 }
 
+template <typename T>
+bool TryCast(const py::object stuf, T &value) {
+	try {
+		value = stuf.cast<T>();
+		return true;
+	} catch (py::cast_error) {
+		return false;
+	}
+}
+
+template <typename T>
+T Cast(const py::object obj) {
+	return obj.cast<T>();
+}
+
+//! 'count' is the amount of rows in the 'out' vector
+//! offset is the current row number within this vector
+void ScanPandasObject(PandasColumnBindData &bind_data, PyObject *object, idx_t offset, Vector &out) {
+
+	// handle None
+	if (object == Py_None) {
+		auto &validity = FlatVector::Validity(out);
+		validity.SetInvalid(offset);
+		return;
+	}
+
+	auto val = TransformPythonValue(object, out.GetType());
+	// Check if the Value type is accepted for the LogicalType of Vector
+	out.SetValue(offset, val);
+}
+
+static void VerifyMapConstraints(Vector &vec, idx_t count) {
+	auto invalid_reason = CheckMapValidity(vec, count);
+	switch (invalid_reason) {
+	case MapInvalidReason::VALID:
+		return;
+	case MapInvalidReason::DUPLICATE_KEY:
+		throw std::runtime_error("Dict->Map conversion failed because 'key' list contains duplicates");
```

Comment:
One of the biggest issues with the python client at present is the widespread use of `runtime_error` within the codebase (which is mapped straight to the python  `RuntimeError`), as it means all exception handling requires the use of string matching on the exception message in duckdb-engine and I imagine other consumers of the api. Would it be possible to fix this before the pr is merged? Ideally there would be a root "DuckDBException" that could be cleanly exposed via pybind11

Reply:
I would say this is out of scope for this PR and should be fixed in a separate one